dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractAnalyticsService.java

```diff
@@ -264,33 +310,62 @@ protected void addMetadata( EventQueryParams params, Grid grid )
         }
 
         params.getItemLegends().stream()
-            .filter( legend -> legend != null )
+            .filter( Objects::nonNull )
             .forEach( legend -> metadataItemMap.put( legend.getUid(),
                 new MetadataItem( legend.getDisplayName(), includeDetails ? legend.getUid() : null,
                     legend.getCode() ) ) );
 
-        params.getItemOptions().stream()
-            .filter( option -> option != null )
-            .forEach( option -> metadataItemMap.put( option.getUid(),
-                new MetadataItem( option.getDisplayName(), includeDetails ? option.getUid() : null,
-                    option.getCode() ) ) );
+        addMetadataItems( metadataItemMap, params, itemOptions );
 
         params.getItemsAndItemFilters().stream()
-            .filter( item -> item != null )
+            .filter( Objects::nonNull )
             .forEach( item -> metadataItemMap.put( item.getItemId(),
                 new MetadataItem( item.getItem().getDisplayName(), includeDetails ? item.getItem() : null ) ) );
 
         return metadataItemMap;
     }
 
+    /**
+     * Add into the MetadataItemMap itemOptions
+     *
+     * @param metadataItemMap MetadataItemMap.
+     * @param params EventQueryParams.
+     * @param itemOptions itemOtion list.
+     */
+    private void addMetadataItems( final Map<String, MetadataItem> metadataItemMap, final EventQueryParams params,
+        final List<Option> itemOptions )
+    {
+        boolean includeDetails = params.isIncludeMetadataDetails();
+
+        if ( !params.isSkipData() )
+        {
+            itemOptions.forEach( option -> metadataItemMap.put( option.getUid(),
+                new MetadataItem( option.getDisplayName(), includeDetails ? option.getUid() : null,
+                    option.getCode() ) ) );
+        }
+        else
+        {
+            params.getItemOptions().stream()
```

Comment:
Could you add a comment here just to give a quick overview of what's going on in this filtering and the reason for it?

Reply:
done